doc/nat/nat_design_spec.md

```diff
@@ -0,0 +1,1422 @@
+The conntrack entries in the kernel and the APP_DB entries are aged out eventually by the inactivity period timeouts since there will not be any NAT traffic crossing over the deleted IP interfaces.
+
+#### 3.3.1.3 Interface link events
+When the outside NAT zone interface's link goes down, the iptables NAT rules, the conntrack entries and the APP_DB dynamic entries are not removed immediately.
+
+This is done for the reasons below:
+
+	- The temporary link flap events happening on the NAT outside interface should not
+	  affect the NAT translation entries (corresponding to the end to end TCP/UDP sessions).
+	- To avoid performance issues during link flaps.
+	- The dynamic entries are aged out eventually due to inactivity timeout in case
+	  the NAT outside interface is link down for more than the timeout period.
+
+### 3.3.2 Natsync daemon
+
+NatSyncd listens to the conntrack netlink notification events from the kernel for the creation, update and removal events of the connections in the conntrack table.
+Once the notifications for SNAT and DNAT entries are received, they are pushed into the NAT_TABLE/NAPT_TABLE/NAPT_TWICE_TABLE in the APP_DB to be consumed by the NatOrch Agent. More details in the section 3.4.
+
+### 3.3.3 NatOrch Agent
+NAT feature is disabled by default.
+When the NAT feature is enabled, NatOrch enables the feature and the NAT miss trap actions in the hardware.
+The SNAT or DNAT miss packets are rate limited to CPU @ 600pps.
+
+NatOrch is responsible for the following activities:
+
+   - Listens on the notifications in the NAT, NAPT, NAPT_TWICE tables in the APP_DB, picks the notifications, translates them to SAI objects and pushes them to ASIC_DB to be consumed by Syncd. For every single NAT/NAPT entry, NatOrch pushes 1 SNAT entry and 1 DNAT entry (for inbound and outbound directions) to the ASIC_DB via sairedis.
+   - Monitors the translation activity status every sampling period for each of the NAT/NAPT entries in the hardware. Sampling period is chosen to be a optimal value of 60 seconds. If the NAPT entry is inactive for the duration of the NAT timeout period, the entry is removed. Static NAT/NAPT entries are not monitored for the inactivity. They have to be unconfigured explicitly to be removed from the hardware.
+
+### 3.3.3.1 Interaction with NeighOrch
+NatOrch registers as observer with the NeighOrch so that the DNAT entry and the corresponding SNAT entry can be pushed to ASIC_DB by NatOrch, only if the corresponding inside zone neighbor's entry is created in NeighOrch.
+When the Neighbor entry is deleted in NeighOrch, the corresponding DNAT and SNAT entries are removed from the hardware.
+ 
+### 3.3.3.2 Aging inactive entries
+Since the active data traffic for the TCP/UDP connections are not flowing through the kernel, the timing out of the connection tracking entries is the responsibility of the application.
```

Comment:
What is the design to make kernel vs ASIC conntrack NAT entries are in sync w.r.t conntrack entries timeout and refresh?

Reply:
NAT application periodically queries the activity of the NAT flows every 30 seconds.
For every active NAT flow, the corresponding conntrack entry's timeout value is reset in the kernel (say set the timeout to a bigger value 3600 secs).
For the inactive NAT flows (that are inactive for the configurable timeout periods of nat-tcp-timeout, nat-udp-timeout), the corresponding connctrack entry is deleted from the kernel by the NatOrch. The delete notification is picked by NatSync and removes the entry from APP_DB and further down it gets deleted from the ASIC.